### PR TITLE
Some more DAS core interfaces and some draft implementations 

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

Outline some more interfaces: 
- `DataColumnReqResp` to abstract over network Req/Resp implementation 
- `DataColumnPeerSearcher` to abstract over `ConnectionManager` + `PeerSelectionStrategy`
- `DataColumnManager` which 
  - notifies on peer connect/disconnect 
  - can ban peers returning invalid req/resp responses (this could probably fit into existing peers scoring classes) 
- `DataAvailabilitySampler`: minimalistic for now 

Add simple implementations of`DataColumnSidecarRetriever` and `DataAvailabilitySampler` interfaces

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/ValidatingDataColumnReqResp.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  @Override
+  public SafeFuture<DataColumnSidecar> requestDataColumnSidecar(
+      UInt256 nodeId, DataColumnIdentifier columnIdentifier) {
+    return reqResp
+        .requestDataColumnSidecar(nodeId, columnIdentifier)
+        .thenCompose(
+            sidecar ->
+                validator
+                    .validate(sidecar)
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/ValidatingDataColumnReqResp.java:45]: nit: we could validate that it matches columnIdentifier. Not important for prototype but easy to miss and forget

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    this.random = random;
+    this.samplesPerSlot = samplesPerSlot;
+    this.extBlobColumnCount = extBlobColumnCount;
+  }
+
+  private Collection<UInt64> getColumnsForSampling() {
+    List<UInt64> allColumnIndexes =
+        Stream.iterate(0, (idx) -> idx < extBlobColumnCount, (idx) -> idx + 1)
+            .map(UInt64::valueOf)
+            .collect(Collectors.toList());
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java]: nit: we have toList() without collect in modern Java

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    ;
+
+    List<UInt64> selectedIndexes = new ArrayList<>();
+    for (int i = 0; i < samplesPerSlot; i++) {
+      int idx = random.nextInt(allColumnIndexes.size());
+      UInt64 columnIndex = allColumnIndexes.remove(idx);
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java:54]: nit: not the best collection for this

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  @SuppressWarnings("unused")
+  private void reqRespCompleted(
+      RetrieveRequest request, DataColumnSidecar maybeResult, Throwable maybeError) {
+    if (maybeResult != null) {
+      synchronized (this) {
+        pendingRequests.remove(request.columnId);
+      }
+      request.peerRequest.dispose();
+    } else {
+      request.activeRpcRequest = null;
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java:143]: shouldn't we clean it in either result?
Also maybe it should be a part of `dispose()`?

**Review by zilm13 — APPROVED** [review]
LGTM

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/ValidatingDataColumnReqResp.java (hunk 1 above)
Comment (Nashatyrev) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/ValidatingDataColumnReqResp.java:45]: Good point. But I believe we would do this on a lower lever when batching/unbatching requests

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java (hunk 2 above)
Comment (Nashatyrev) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java]: It turned out that `toList()` returns unmodifiable instance, but we need a mutable list

Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/SimpleDataAvailabilitySampler.java]: ahhh, right, my bad

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java (hunk 4 above)
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java:143]: ahh we don't need to reuse it, ok, got it.

**Pull request merged**